src/librustc_target/spec/apple_base.rs

```diff
@@ -51,3 +51,63 @@ pub fn macos_llvm_target(arch: &str) -> String {
     let (major, minor) = macos_deployment_target();
     format!("{}-apple-macosx{}.{}.0", arch, major, minor)
 }
+
+#[cfg(target_os = "macos")]
+pub fn sysroot(sdk: &str) -> Result<Option<String>, String> {
+    // Like Clang, allow the `SDKROOT` environment variable used by Xcode to define the sysroot.
+    if let Some(sdk_root) = env::var("SDKROOT").ok() {
+        let actual_sdk_path = sdk_path(sdk)?;
+        let sdk_root_p = Path::new(&sdk_root);
+        // Ignore `SDKROOT` if it's not a valid path.
+        if !sdk_root_p.is_absolute() || sdk_root_p == Path::new("/") || !sdk_root_p.exists() {
+            return Ok(Some(actual_sdk_path));
+        }
+        // Ignore `SDKROOT` if it's clearly set for the wrong platform, which may occur when we're
+        // compiling a custom build script while targeting iOS for example.
+        return Ok(Some(match sdk {
+            "iphoneos" if sdk_root.contains("iPhoneSimulator.platform")
+                || sdk_root.contains("MacOSX.platform") => actual_sdk_path,
+            "iphonesimulator" if sdk_root.contains("iPhoneOS.platform")
+                || sdk_root.contains("MacOSX.platform") => actual_sdk_path,
+            "macosx" | "macosx10.15" if sdk_root.contains("iPhoneOS.platform")
+                || sdk_root.contains("iPhoneSimulator.platform") => actual_sdk_path,
+            _ => sdk_root,
+        }))
+    }
+    Ok(None)
+}
+
+// `xcrun` is only available on macOS.
+#[cfg(not(target_os = "macos"))]
```

Comment:
None of the changes here are using macOS-specific APIs in the binary, so could `#[cfg]` be avoided here and instead `if cfg!(...)` used instead? I think we'll want to gracefully handle errors from `xcrun` regardless